pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+        {
+            const auto prepared_proof = viewchange_message.prepared_proofs(i);
+
+            //  --- if we have not already created a new preprepare for an operation with that sequence number
+
+            //  ---- Create a new preprepare for that operation using its original sequence number
+            //       and request hash, but using the new view number
+
+            bzn_envelope preprepare_envelope;
+            preprepare_envelope.ParseFromString(prepared_proof.pre_prepare());
+
+            pbft_msg pre_prepare;
+            pre_prepare.ParseFromString(preprepare_envelope.pbft());
+            pre_prepare.set_view(new_view);
+
+            if (pre_prepare.sequence() < this->latest_stable_checkpoint().first)
+            {
+                continue;
+            }
+            pre_prepares[pre_prepare.sequence()] = this->make_signed_envelope(pre_prepare.SerializeAsString());
+        }
+
+        //  - add each preprepare created this way to the new-view message
+    }
+
+    //  - for each sequence number between the base-sequence number (the last stable checkpoint)
+    //    and the highest sequence number for which a new preprepare was created
+
+    // Fill in missing messages with no ops
+    this->fill_in_missing_pre_prepares(pre_prepares);
+
+    //  - add each preprepare created this way to the new-view message
+    // std::set<std::string> view_change_messages
+    // viewchange_messages
+
+    return this->make_newview(new_view, viewchange_messages, pre_prepares);
+}
+
+bzn_envelope
+pbft::make_signed_envelope(std::string serialized_pbft_message)
+{
+    bzn_envelope envelope;
+    envelope.set_pbft(serialized_pbft_message);
+    envelope.set_sender(this->get_uuid());
+    if (!this->crypto->sign(envelope))
+    {
+        LOG (error) << "make_signed_envelope - failed to sign envelope";
+    }
+
+    return envelope;
+}
+
+/**
+ *
+ * @param msg
+ */
+void
+pbft::save_checkpoint(const pbft_msg& msg)
+{
+    pbft_msg message;
+    for (int i{0}; i < msg.checkpoint_messages_size(); ++i)
+    {
+        const auto checkpoint_msg = msg.checkpoint_messages(i);
+        bzn_envelope original_checkpoint;
+        if (original_checkpoint.ParseFromString(checkpoint_msg)
+            || !this->crypto->verify(original_checkpoint) // TODO: consider removing this - check with debugger
+            || !message.ParseFromString(original_checkpoint.pbft()))
+        {
+            continue;
+        }
+        this->handle_checkpoint(message, original_checkpoint);
+    }
+}
+
+void
+pbft::replica_broadcasts_viewchange(const pbft_msg& msg)
+{
+    // replica sends VIEWCHANGE message
+    pbft_msg viewchange_message{
+            make_viewchange(
+                    msg.view()
+                    , this->latest_stable_checkpoint().first
+                    , this->stable_checkpoint_proof
+                    , this->prepared_operations_since_last_checkpoint())
+    };
+
+    this->view_is_valid = false;
```

Comment:
don't we already set this in handle_failure?

Reply:
Yes, removed.